### PR TITLE
banking_stage: Remove old Tpu leader rotation shutdown mechanism

### PR DESCRIPTION
**Pull request by mvines**
When the PohService was introduced to signal leader rotation once it achieved `max_tick_height` [1], the previous mechanism of having the first banking_stage thread that achieves `max_tick_height` emit the signal was not removed and we acquired some dead code, which is removed in this PR

[1] https://github.com/solana-labs/solana/blob/e7df3cfe22c9f5e8f1dbf7014fd374910028ef63/src/poh_service.rs#L94-L97

### CODE REVIEW

**Comment by carllin** [comment]
That banking stage check is still needed for the scenario where: 

1) The PoHService registers and sends the last tick here: https://github.com/solana-labs/solana/blob/master/src/poh_recorder.rs#L52. It doesn't error out with MaxHeightReached until the next time tick() is called

2) The banking stage thread tries to record a tx and gets the MaxHeightReached error

3)  The banking stage thread sets the exit signal causing PohService to exit before it can signal leader rotation.

**Comment by mvines** [comment]
No, I don't think that's true anymore.  Maybe it was in the past.  The `Error::PohRecorderError(PohRecorderError::MaxHeightReached)` case never occurs, as the bank now sends `Error::BankError(BankError::MaxHeightReached)` which prior to this patch was hitting the default `error!("solana-banking-stage-tx {:?}", e),` and doing nothing beyond logging an `error!` 🙃 

All tests are green with this patch too, so either we have a big test coverage hole or it truly is dead code.

**Comment by carllin** [comment]
Ah yeah, so that means we need all the logic currently in the case for Error::PohRecorderError(PohRecorderError::MaxHeightReached) in Error::BankError(BankError::MaxHeightReached). Otherwise, the exit flag set from the banking thread on breaking out of the loop can cause PoHRecorder to exit without signaling leader rotation

**Comment by mvines** [comment]
Hmm, ok.  That's scary we don't have a test for this.  I'll try write one to prove the code I removed actually matters, or not

**Comment by mvines** [comment]
@carllin wdyt about just changing the `Error::BankError(BankError::MaxHeightReached)` handling in banking_stage to do nothing.  This effective is the case at the tip of master where leader rotation is generally working.  In this case, PoHRecorder is the sole authority on initiating a leader rotation and all the banking stage threads behave the same -- they either stop when they hit max height, eventually bail with a `BankingStageReturnType::ChannelDisconnected` (which this literally how master currently behaves)

**Comment by mvines** [comment]
176ce9069232c0a8a167c7dca1b5be589b790539 is what I proposed in my previous comment

**Comment by carllin** [comment]
@mvines, yup I think that's totally fine. Adding @sagar-solana, to double check there's no edge cases we're missing

**Comment by sagar-solana** [comment]
@mvines Yeah i like the thought of banking stage having nothing to do with rotation control. Although, I added the rotation signal here because I was seeing the tpu not signaling rotation anx exiting anyway. It would cause the whole network to stall. What ended up happening was that banking stage would exit and so would poh recorder but nothing send a message on the channel.

Before the persistent tvu pr, handle role transition would join on the tpu, and so it didn't matter how an exit happened. It would just assume rotation and restart. I'm not sure that is handled here. I guess the senders will all be dropped and that _should_ trigger something. It's a bit racey so just wanted to be sure that is something we won't run into again.

**Comment by mvines** [comment]
🛎 @carllin, @sagar-solana - PR is ready for r?

File: src/banking_stage.rs [hunk 1]
```diff
@@ -32,7 +32,6 @@ use sys_info;
 pub enum BankingStageReturnType {
     LeaderRotation(u64),
     ChannelDisconnected,
-    RecordFailure,
```
Comment (mvines) [inline src/banking_stage.rs]: Nobody uses the `RecordFailure` enum, kill it

File: src/banking_stage.rs [hunk 2]
```diff
@@ -112,31 +106,28 @@ impl BankingStage {
...
                                 }
                             }
                             if thread_banking_exit.load(Ordering::Relaxed) {
                                 break None;
                             }
                         };
-                        thread_banking_exit.store(true, Ordering::Relaxed);
+
+                        // Signal exit only on "Some" error
+                        if return_result.is_some() {
```
Comment (mvines) [inline src/banking_stage.rs]: Here's where we avoid shutting down the rest of the Tpu if/when the Bank hits MaxHeightReached before the PohRecorder does.

Comment (sagar-solana) [reply src/banking_stage.rs]: Yup okay makes sense.

**Pull request merged**